src/cons_list.rs

```diff
@@ -78,31 +70,70 @@ impl<T, Ts: ConsListT<T>> ConsList<T, Cons<T, Ts>> {
     }
 }
 
-pub struct FullyConsumeOnDrop<I: Iterator>(I);
+impl<T, Ts: ConsListT<T>> IntoIterator for ConsList<T, Ts> {
+    type Item = T;
+    type IntoIter = Iter<T, Ts>;
 
-impl<I: Iterator> Iterator for FullyConsumeOnDrop<I> {
-    type Item = I::Item;
+    fn into_iter(self) -> Self::IntoIter {
+        let ConsList { list, .. } = self;
+        Iter::new(list)
+    }
+}
+
+pub struct Iter<T, Ts: ConsListT<T>> {
+    list: ManuallyDrop<Ts>,
+    alive: Range<usize>,
+    marker: PhantomData<T>,
+}
+
+impl<T, Ts: ConsListT<T>> Iter<T, Ts> {
+    fn new(list: Ts) -> Self {
+        Iter {
+            list: ManuallyDrop::new(list),
+            alive: 0..Ts::LEN,
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<T, Ts: ConsListT<T>> Iterator for Iter<T, Ts> {
+    type Item = T;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.0.next()
+        self.alive
+            .next()
+            .map(|idx| unsafe { self.list.take_unchecked(idx) })
     }
 
     fn size_hint(&self) -> (usize, Option<usize>) {
-        self.0.size_hint()
+        let len = self.len();
+        (len, Some(len))
+    }
+
+    fn count(self) -> usize {
```

Comment:
This won't actually be faster than the default impl since we have to consume the list on drop

Reply:
this Iterator impl is a trimmed down version of what is implemented for Array. So I will defer to the std library on this one